### ISSUE TITLE
One failing rows make fail the entire batch

### ISSUE THREAD

**Issue by alepuccetti** [logstash-plugins/logstash-output-google_bigquery#44]
While I am doing some tests I noticed that if I have bad rows (e.i. malformed timestamps) the entire batch fails to load even if only one row is bad.

Is this a known issue, desired behavior, bigquery streaming API issue?

**Comment by alepuccetti** [comment]
Hi @josephlewis42,

> This is a known issue, I have a fix for it in #40 that will let you upload as much as possible, keeping only the rejected rows to a file. That PR has sat there because there wasn't anyone around to test it. Elastic likes to see a PR done, reviewed, then tested by a third-party community member before doing a release.

That's great

I could fit this testing in my spare time. I use docker and I am not doing much java. It would be super easy if I can install the plugin there.

Cheers,

**Comment by josephlewis42** [comment]
Hi @alepuccetti,

I built a release gem and put it here so you can try it: https://github.com/josephlewis42/logstash-output-google_bigquery/releases/tag/4.1.0-pre

You'll want to set `skip_invalid_rows` to true in your configuration to get the new feature.

Here are the docs I wrote for it:

```
===== `skip_invalid_rows`
 added[4.1.0]
   * Value type is <<boolean,boolean>>
  * Default value is `false`
 Insert all valid rows of a request, even if invalid rows exist.
The default value is false, which causes the entire request to fail if any invalid rows exist.

```

**Comment by alepuccetti** [comment]
Hi @josephlewis42,

I am finding some trouble to install the pre-release. I am using docker so I run this:

```
RUN curl https://github.com/josephlewis42/logstash-output-google_bigquery/releases/download/4.1.0-pre/logstash-output-google_bigquery-4.1.0-java.gem --output logstash-output-google_bigquery-4.1.0-java.gem && \
bin/logstash-plugin install --no-verify logstash-output-google_bigquery-4.1.0-java.gem
```

but I get this error:

```
Gem::Package::FormatError: package metadata is missing in logstash-output-google_bigquery-4.1.0-java.gem
                         verify_files at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package.rb:599
                      block in verify at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package.rb:524
                                  new at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package/tar_reader.rb:29
                      block in verify at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package.rb:521
                                 open at org/jruby/RubyIO.java:1156
                                 open at org/jruby/RubyKernel.java:306
                         with_read_io at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package/file_source.rb:30
                               verify at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package.rb:520
                        extract_files at /usr/share/logstash/vendor/jruby/lib/ruby/stdlib/rubygems/package.rb:335
                               unpack at /usr/share/logstash/lib/bootstrap/rubygems.rb:34
  block in extract_local_gems_plugins at /usr/share/logstash/lib/pluginmanager/install.rb:171
                              collect at org/jruby/RubyArray.java:2481
           extract_local_gems_plugins at /usr/share/logstash/lib/pluginmanager/install.rb:161
                              execute at /usr/share/logstash/lib/pluginmanager/install.rb:52
                                  run at /usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:67
                              execute at /usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/subcommand/execution.rb:11
                                  run at /usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:67
                                  run at /usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:132
                               <main> at /usr/share/logstash/lib/pluginmanager/main.rb:48
```

Do you have any tips?

**Comment by josephlewis42** [comment]
Hi @alepuccetti,

Would you try adding a -L to that `curl` command? I think GitHub might be returning an HTML page with a JavaScript redirect rather than the gem:

```
RUN curl -L https://github.com/josephlewis42/logstash-output-google_bigquery/releases/download/4.1.0-pre/logstash-output-google_bigquery-4.1.0-java.gem --output logstash-output-google_bigquery-4.1.0-java.gem && \
bin/logstash-plugin install --no-verify logstash-output-google_bigquery-4.1.0-java.gem
```

**Comment by alepuccetti** [comment]
Hi @josephlewis42,

Thanks for the tip. I tested and it worked like a charm I am still running it in my test environment. I will live it running for the day and report if there is any problem but I started it an hour ago and no problem so far.

It is logging the failing lines and only the failing lines. It does not fail the entire batch and all the lines seem to be in the DB.

Great job 🎉 .

**Comment by alepuccetti** [comment]
@josephlewis42 I am running your patch for about 30 hours now. And everything looks good.

**Comment by josephlewis42** [comment]
Hi @alepuccetti I got access to publish and the fix is merged and should be out now as version 4.1.0
